internal/twirptest/service_test.go

```diff
@@ -1527,3 +1527,150 @@ func TestPanicFlushing(t *testing.T) {
+	}
+}
+
+func TestDeadlineExceededError(t *testing.T) {
+	// test case validate context deadline exceed error handling
+	var s Haberdasher
+	hooks := &twirp.ServerHooks{Error: func(ctx context.Context, err twirp.Error) context.Context {
+		if ctxErr := twirp.DeadlineExceeded; err.Code() != ctxErr {
+			t.Errorf("twirp ErrorCode expected to be %q, but found %q", ctxErr, err.Code())
+		}
+		if ctxErr := context.DeadlineExceeded.Error(); err.Msg() != ctxErr {
+			t.Errorf("twirp client err has unexpected message %q, want %q", err.Msg(), ctxErr)
+		}
+		return ctx
+	}}
+	server := NewHaberdasherServer(s, hooks)
+
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 1*time.Millisecond)
+	defer cancel()
+	// Make a request, that will endpoint method
+	req, _ := http.NewRequest(http.MethodPost, "http://testing:8080/twirp/twirp.internal.twirptest.Haberdasher/MakeHat", errReader{})
+	req.Header.Set("Accept", "application/protobuf")
+	req.Header.Set("Content-Type", "application/protobuf")
+	// Associate the cancellable context we just created to the request
+	req = req.WithContext(ctx)
+
+	// wait for deadline exceed timeout in context object
+	select {
+	case <-ctx.Done():
+		// ctx.Err() will have context.DeadlineExceeded
+	case <-time.After(1 * time.Second):
+		t.Fatalf("Waited too long for DeadlineExceeded ")
+	}
+	w := httptest.NewRecorder()
+
+	server.ServeHTTP(w, req)
+
+	// validate error code
+	expectedErrCode := twirp.ServerHTTPStatusFromErrorCode(twirp.DeadlineExceeded)
+	if w.Code != expectedErrCode {
+		t.Errorf("twirp ErrorCode expected to be %q, but found %q", expectedErrCode, w.Code)
+	}
+
+	// validate error message
+	respBytes, err := ioutil.ReadAll(w.Body)
+	if err != nil {
+		t.Fatalf("Could not even read bytes from response: %q", err.Error())
+	}
+
+	expectedErrMessage := "context deadline exceeded"
+	if !strings.Contains(string(respBytes), expectedErrMessage) {
+		t.Errorf("twirp client err has unexpected message %q, want %q", string(respBytes), expectedErrMessage)
+	}
+}
+
+func TestFailedToReadRequestBodyError(t *testing.T) {
+	// test case validate reqest body read error handling
+	var s Haberdasher
+	hooks := &twirp.ServerHooks{Error: func(ctx context.Context, err twirp.Error) context.Context {
+		if ctxErr := twirp.Malformed; err.Code() != ctxErr {
+			t.Errorf("twirp ErrorCode expected to be %q, but found %q", ctxErr, err.Code())
+		}
+		if ctxErr := "failed to read request body"; err.Msg() != ctxErr {
+			t.Errorf("twirp client err has unexpected message %q, want %q", err.Msg(), ctxErr)
+		}
+		return ctx
+	}}
+	server := NewHaberdasherServer(s, hooks)
+	ctx := context.Background()
+
+	// Make a request, that will call endpoint method
+	req, _ := http.NewRequest(http.MethodPost, "http://testing:8080/twirp/twirp.internal.twirptest.Haberdasher/MakeHat", errReader{})
+	req.Header.Set("Accept", "application/protobuf")
+	req.Header.Set("Content-Type", "application/protobuf")
```

Comment:
The tests should also cover behavior of the JSON client.

Reply:
Thanks. I think error handling is missing for JSON client also and I end up adding that too. Please let me know if I am wrong.
https://github.com/twitchtv/twirp/pull/294/files#diff-ca331af0419d8cb15dfaf494e87bc6fac163b0f5a897b9927cd1eff3b422b2a8R1272-R1278